File: notebooks/anywidget_jax_callback.py
# %%
# Example of using JAX's io_callback to send messages to an AnyWidget instance.
# The widget responds, but python won't receive the message until after the
# current cell has evaluated.

from jax.experimental import io_callback
import jax
import jax.numpy as jnp
import anywidget

# %%


class Widget(anywidget.AnyWidget):
    _esm = "anywidget_jax_callback.js"

    def __init__(self):
        super().__init__()

    @anywidget.experimental.command
    def _receive_message(self, msg, buffers):
        print(f"Received message from js: {msg}")
        return ["ok", []]


w = Widget()
w
# %%


def effectful_fn(x):
    print(f"Performing an effect with {x}")
    w.send({"kind": "genstudio", "content": x.tolist()})
    return x


@jax.jit
def numpy_random_like(x):
    return io_callback(effectful_fn, x, x)


jax.vmap(numpy_random_like)(jnp.zeros(5))

File: notebooks/anywidget_jax_callback.js
import {createRender, useModel, useExperimental} from "https://esm.sh/@anywidget/react@0.0.7"
import * as React from 'https://esm.sh/react@18.3.1'
import htm from 'https://esm.sh/htm@3.1.1'

const {useState, useEffect} = React
const html = htm.bind(React.createElement)

  const useCustomMessages = (model) => {
      const [messages, setMessages] = useState([])
      const handleMessage = (message, info) => {
        if (message.kind !== 'genstudio') {
          return;
        }
        setMessages((ms) => [...ms, message.content]);
      };
      useEffect(() => {
        model.on("msg:custom", handleMessage)
        return () => model.off("msg:custom", handleMessage)
      }, [model])
      return messages
  }

  const App = () => {
    const model = useModel()
    const {invoke} = useExperimental()
    const messages = useCustomMessages(model)
        useEffect(() => {
            if (messages.length > 0) {
                // Send a response to python
                invoke("_receive_message", messages[messages.length - 1], [])
            }
            },
        [messages.join()])
        return html`<div>
                    Messages: ${messages.join(', ')}
                </div>`
    }
  export default {render: createRender(App)}
